query/src/main/scala/filodb/query/exec/SetOperatorExec.scala

```diff
@@ -187,59 +188,175 @@ final case class SetOperatorExec(queryContext: QueryContext,
 
-    val lhsResult = new mutable.HashMap[Map[Utf8Str, Utf8Str], ArrayBuffer[RangeVector]]()
-    val rhsResult = new mutable.HashMap[Map[Utf8Str, Utf8Str], ArrayBuffer[RangeVector]]()
+    val lhsResult = groupAndStitchRangeVectors(lhsRvs)
+    val rhsResult = groupAndStitchRangeVectors(rhsRvs)
+    // All lhsResults should be considered
+    // All rhsResults whose joinKey is not present in lhsResults should be considered
+    // While iterating all lhs rangevectors for a time instant where none of the lhs range vectors are having a value
+    // all RHS range vectors for that time instant having a matching join key will have a value in o/p
+
+    //lhsResult.valuesIterator.map(_.toIterator).flatten ++ rhsResult.valuesIterator.map(_.toIterator).flatten
+
+    // Create a Map of join key as key and value is a two tuple where _1 is all LHS Rvs for the join key and _2
+    // are all RHS Rvs for that same join key
+    val lRvs = lhsResult.foldLeft(Map.empty[Map[Utf8Str, Utf8Str], (Seq[RangeVector], Seq[RangeVector])]) {
+      case (acc, (jk, rvs)) => acc + (jk -> ((rvs, Seq.empty[RangeVector])))
+    }
+
+    val joinedRvs = rhsResult.foldLeft(lRvs) {
+      case (acc, (jk, rvs))  =>
+        val newVal = acc.get(jk).map {
+            // RHS will be Nil
+          case (lhs, rhs) => (lhs, rvs)
+        }.getOrElse((Nil, rvs))
+        acc + (jk -> newVal)
+    }
+
+    case class OrSetOpIteratorWrapper(lhs: Seq[RangeVector], rhs: Seq[RangeVector]) {
+      // Has a side effect where LHS are consumed first and a bit is set to indicate the time instant where there
+      // is at least one non Nan value. the rhs values are then iterated over and emitted only where the given timestep
+      // has no value in any of the LHS timeseries
+      assert(outputRvRange.isDefined, "Expected to see a non null outputRvRange")
+      val (start, end, step) = (outputRvRange.get.startMs, outputRvRange.get.endMs, outputRvRange.get.stepMs)
+      // 64 bits in one word (long), 1 bit per step
+      assert(start == end || step > 0, "Expected to see a non 0 step size when start and end are not same")
+      val wordsRequired = if (start == end) 1 else ((((end - start) / step) >> 6) + 1).toInt
+      val bitSet = new mutable.BitSet(wordsRequired)
+
+      def joined(): Seq[RangeVector] = {
+        val lhsRvs: Seq[RangeVector] = lhs.map(x => {
+          var idx = 0
+          val rows = x.rows()
+          val rowsCursor: RangeVectorCursor = new RangeVectorCursor {
+            val cur = new TransientRow()
+
+            override def hasNext: Boolean = rows.hasNext
+
+            override def next(): RowReader = {
+              val reader = rows.next()
+              // TODO: What is the expected behavior for Histograms?
+              val value = reader.getDouble(1)
+              val res = if (value.isNaN) {
+                Double.NaN
+              } else {
+                bitSet += idx
```

Comment:
Since we only set the bit when the value is not NaN, the RHS value will be returned. But Prometheus would return the NaN: http://demo.robustperception.io:9090/graph?g0.expr=vector(Nan)%20or%20sum(up)&g0.tab=1&g0.stacked=0&g0.range_input=1h

Reply:
Can't distinguish NaN for a missing value vs a legitimate value. I guess supporting this if required will need more changes.